salt/orch/kubernetes.sls

```diff
@@ -1,9 +1,41 @@
+{%- set default_batch = 5 %}
+
+{# machine IDs that have the master roles assigned #}
 {%- set masters = salt.saltutil.runner('mine.get', tgt='G@roles:kube-master', fun='network.interfaces', tgt_type='compound').keys() %}
 {%- set super_master = masters|first %}
 
-{%- set default_batch = 5 %}
+{# machine IDs in the cluster that are currently etcd servers #}
+{%- set current_etcd_members = salt.saltutil.runner('mine.get', tgt='G@roles:etcd', fun='network.interfaces', tgt_type='compound').keys() %}
+{%- set num_current_etcd_members = current_etcd_members|length %}
+
+{# the number of etcd masters that should be in the cluster #}
+{%- set num_etcd_members = salt.caasp_etcd.get_cluster_size() %}
+{# ... and the number we are missing #}
+{%- set num_additional_etcd_members = num_etcd_members - num_current_etcd_members %}
+
+{%- set new_etcd_members = [] %}
+
+{%- if num_additional_etcd_members > 0 %}
+
+  {%- set masters_no_etcd = salt.saltutil.runner('mine.get', tgt='G@roles:kube-master and not G@roles:etcd', fun='network.interfaces', tgt_type='compound').keys() %}
+
+  {# get k8s masters until we complete the etcd cluster #}
+  {%- set masters_and_etcd = masters_no_etcd[:num_additional_etcd_members] %}
+  {%- set new_etcd_members = new_etcd_members + masters_and_etcd %}
+  {%- set num_additional_etcd_members = num_additional_etcd_members - masters_and_etcd|length %}
+
+  {# if we have run out of k8s masters and we do not have #}
+  {# enough etcd members, go for the k8s workers too... #}
+  {%- if num_additional_etcd_members > 0 %}
+    {%- set workers_no_etcd = salt.saltutil.runner('mine.get', tgt='G@roles:kube-minion and not G@roles:etcd', fun='network.interfaces', tgt_type='compound').keys() %}
 
-{%- set num_etcd_masters = salt.caasp_etcd.get_cluster_size() %}
+    {%- set workers_and_etcd = workers_no_etcd[:num_additional_etcd_members] %}
+    {%- set new_etcd_members = new_etcd_members + workers_and_etcd %}
+    {%- set num_additional_etcd_members = num_additional_etcd_members - workers_and_etcd|length %}
+
+    {# TODO: if num_additional_etcd_members > 0: fail/message/something... #}
+  {%- endif %}
+{%- endif %}
```

Comment:
I think this logic is misplaced here, and should probably be moved to Velum, where velum sets the master/minion roles already.

Reply:
We have been calculating the optimal number of etcd nodes for a long time. This is just doing the same, but going a bit further and assigning the `etcd` role to those nodes. At the same time, this mechanism does not conflict with Velum: if someone sets the `etcd` roles to nodes, it will respect that...